### PR TITLE
Create standalone e2e test image

### PR DESCRIPTION
**Pull request by chris-crone**
**- What I did**
- Build a standalone e2e test image that can be run against a cluster
- Add `docker stack` e2e tests for Kubernetes

**- How I did it**

Test modifications:
- Skip tests that require engine configuration modifications when running against a remote endpoint.
- Be less strict about CLI outputs to accommodate running the tests against a Swarm classic endpoint.
- Refactor stack tests so that we run the same test against Swarm and Kubernetes.
- Skip Kubernetes tests unless we have a `KUBECONFIG`.
- Add ability to skip tests that manipulate plugins so that this can be used against EE.

e2e image modifications:

- Do not require mounting volumes to run – image is now completely self-contained.

**- How to verify it**
- The existing flow for running e2e tests should work as before: `make -f docker.Makefile test-e2e`
- Run the image against a cluster running swarm: `docker run --rm -v /path/to/certs:/bundle -e DOCKER_HOST=tcp://host_ip:2376 -e DOCKER_CERT_PATH=/bundle -e DOCKER_TLS_VERIFY=1 -e REMOTE_DAEMON=1 -e SKIP_PLUGIN_TESTS=1 docker-cli-e2e`
- Run the image against a cluster running EE: `docker run --rm -v /path/to/certs:/bundle -e DOCKER_HOST=tcp://host_ip:443 -e DOCKER_CERT_PATH=/bundle -e DOCKER_TLS_VERIFY=1 -e REMOTE_DAEMON=1 -e SKIP_PLUGIN_TESTS=1 docker-cli-e2e`

**- Description for the changelog**

N/A testing only.

### CODE REVIEW

File: e2e/stack/remove_test.go [hunk 1]
```diff
@@ -13,45 +14,66 @@ import (
...
+	golden.Assert(t, result.Stdout(),
+		fmt.Sprintf("stack-remove-%s-success.golden", orchestrator))
 }
 
-func deployFullStack(t *testing.T, stackname string) {
+func deployFullStack(t *testing.T, orchestrator, stackname string) {
 	// TODO: this stack should have full options not minimal options
-	result := icmd.RunCommand("docker", "stack", "deploy",
-		"--compose-file=./testdata/full-stack.yml", stackname)
+
```
Comment (mat007) [inline e2e/stack/remove_test.go]: nit: extra line

File: e2e/stack/remove_test.go [hunk 2]
```diff
@@ -13,45 +14,66 @@ import (
...
+	t.Run("Kubernetes", func(t *testing.T) {
+		environment.SkipIfKubernetesDisabled(t)
 
-	result := icmd.RunCommand("docker", "stack", "rm", stackname)
+		testRemove(t, "kubernetes")
+	})
+}
+
+func testRemove(t *testing.T, orchestrator string) {
+	stackname := fmt.Sprintf("test-stack-remove-%s", orchestrator)
```
Comment (mat007) [inline e2e/stack/remove_test.go]: nit: could be `"test-stack-remove-" + orchestrator`

File: e2e/stack/remove_test.go [hunk 3]
```diff
@@ -13,45 +14,66 @@ import (
...
+	deployFullStack(t, orchestrator, stackname)
+	defer cleanupFullStack(t, orchestrator, stackname)
+
+	result := icmd.RunCommand("docker", "--orchestrator", orchestrator,
+		"stack", "rm", stackname)
```
Comment (mat007) [inline e2e/stack/remove_test.go]: nit: why the new line in the middle? (same a few lines down)

File: internal/test/output/output.go [hunk 4]
```diff
@@ -7,15 +7,15 @@ import (
 	"github.com/pkg/errors"
 )
 
-// Assert checks wether the output contains the specified lines
-func Assert(t *testing.T, actual string, expectedLines map[int]func(string) error) {
+// Assert checks output lines at specified locations
+func Assert(t *testing.T, actual string, expectedLines map[int]func(string, bool) error) {
```
Comment (mat007) [inline internal/test/output/output.go]: What is that new `bool` for? Shouldn't we document it?

File: internal/test/output/output.go [hunk 5]
```diff
@@ -24,22 +24,81 @@ func Assert(t *testing.T, actual string, expectedLines map[int]func(string) erro
...
+		found := false
+		var err error
+		for count, line := range toCheck {
+			if err = cmp(line, false); err == nil {
+				toCheck = toCheck[count:]
+				found = true
+				break
+			}
+		}
+		if found == false {
```
Comment (mat007) [inline internal/test/output/output.go]: Did you mean `if !found`? :)

File: internal/test/output/output.go [hunk 6]
```diff
@@ -24,22 +24,81 @@ func Assert(t *testing.T, actual string, expectedLines map[int]func(string) erro
...
+func Prefix(expected string) func(string, bool) error {
+	return func(actual string, exact bool) error {
 		if strings.HasPrefix(actual, expected) {
 			return nil
 		}
-		return errors.Errorf("expected %s to start with %s", actual, expected)
+		if exact {
+			return errors.Errorf("expected %q to start with %q", actual, expected)
+		} else {
+			return errors.Errorf("unable to find line that starts with %q", expected)
```
Comment (mat007) [inline internal/test/output/output.go]: I believe the `else` should be removed:
```
		if exact {
			return errors.Errorf("expected %q to start with %q", actual, expected)
		}
		return errors.Errorf("unable to find line that starts with %q", expected)
```

File: dockerfiles/Dockerfile.e2e [hunk 7]
```diff
@@ -0,0 +1,38 @@
+FROM golang:1.9.5-alpine3.6
```
Comment (vdemeester) [inline dockerfiles/Dockerfile.e2e]: 1.10.2 :angel: :pray:

File: e2e/container/run_test.go [hunk 8]
```diff
@@ -25,6 +28,8 @@ func TestRunAttachedFromRemoteImageAndRemove(t *testing.T) {
 }
 
 func TestRunWithContentTrust(t *testing.T) {
+	environment.SkipIfRemoteDaemon(t)
```
Comment (vdemeester) [inline e2e/container/run_test.go]: nit: Just to look a bit more inline with what we do on other part (in moby or else) : 

```go
skip.If(t, environment.RemoteDaemon())
```

File: e2e/plugin/trust_test.go [hunk 9]
```diff
@@ -19,6 +20,8 @@ import (
 const registryPrefix = "registry:5000"
 
 func TestInstallWithContentTrust(t *testing.T) {
+	environment.SkipIfPluginTestDisabled(t)
```
Comment (vdemeester) [inline e2e/plugin/trust_test.go]: Same as `RemoteDaemon`

File: e2e/container/run_test.go (hunk 8 above)
Comment (chris-crone) [reply e2e/container/run_test.go]: The reason I did it like this was to have the same skip message if we're skipping for the same reason.

I guess I could do `skip.If(t, environment.RemoteDaemon(), environment.RemoteDaemonSkipMessage)`

wdyt?

File: dockerfiles/Dockerfile.e2e (hunk 7 above)
Comment (chris-crone) [reply dockerfiles/Dockerfile.e2e]: Should I bump the alpine version too?

File: e2e/compose-env.yaml [hunk 10]
```diff
@@ -10,12 +10,9 @@ services:
       command: ['--insecure-registry=registry:5000']
 
   notary-server:
-      image: 'notary:server-0.4.2'
+      build:
+        context: ..
```
Comment (thaJeztah) [inline e2e/compose-env.yaml]: Looks like we only need `testdata` from this directory;

```Dockerfile
COPY e2e/testdata/notary/ /fixtures
```

May want to change this to

```
context: ./testdata/
```

And the Dockerfile to;

```
COPY ./notary/ /fixtures/
```

(or a variation thereof)

File: internal/test/environment/testenv.go [hunk 11]
```diff
@@ -14,7 +17,78 @@ func Setup() error {
...
+}
+
+// KubernetesEnabled returns if Kubernetes testing is enabled
+func KubernetesEnabled() bool {
+	return os.Getenv("KUBECONFIG") != ""
+}
+
+// SkipIfKubernetesDisabled skips the test if Kubernetes is not enabled
+func SkipIfKubernetesDisabled(t *testing.T) {
+	skip.If(t, !KubernetesEnabled(), "Kubernetes not enabled")
```
Comment (dnephin) [inline internal/test/environment/testenv.go]: These helpers aren't really necessary. It should already echo the second argument which reads basically the same way as the extra string.

If you keep them I would add `t.Helper()` here so at least it reports the correct line in the skip message.

File: internal/test/output/output.go [hunk 12]
```diff
@@ -24,22 +25,78 @@ func Assert(t *testing.T, actual string, expectedLines map[int]func(string) erro
 	}
 }
 
-// Prefix returns whether if the line has the specified string as prefix
-func Prefix(expected string) func(string) error {
-	return func(actual string) error {
+// AssertContains checks whether the output contains specified lines in order
+func AssertContains(t *testing.T, actual string, expectedLines []func(string, bool) error) {
```
Comment (dnephin) [inline internal/test/output/output.go]: Why is this necessary? Previously it was working with `Assert`.

Comment (chris-crone) [reply internal/test/output/output.go]: I added this function so that the build test would pass against Docker EE which uses a 17.06 engine. With this engine, `COPY` steps have an extra `Removing intermediate container` line. This new function doesn't check exact line numbers but instead checks for desired outputs in order.

**17.06:**
```bash
$ docker build -t copy .
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM alpine
 ---> 3fd9065eaf02
Step 2/2 : COPY something /
 ---> 577b61f8471b
Removing intermediate container 86ec5fc54328
Successfully built 577b61f8471b
Successfully tagged copy:latest
```

**18.06:**
```bash
$ docker build -t copy .
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM alpine
 ---> 3fd9065eaf02
Step 2/2 : COPY something /
 ---> 93321d321367
Successfully built 93321d321367
Successfully tagged copy:latest
```

Comment (vdemeester) [reply internal/test/output/output.go]: > Removing intermediate container 86ec5fc54328

This line disappears ? is it by design ? cc @thaJeztah @simonferquel

File: e2e/testdata/Dockerfile.notary-server [hunk 13]
```diff
@@ -0,0 +1,4 @@
+ARG NOTARY_VERSION=0.5.0
```
Comment (vdemeester) [inline e2e/testdata/Dockerfile.notary-server]: Why `0.5.0` here ? `ARG     NOTARY_VERSION=v0.6.0`, it was `0.6.0` before :stuck_out_tongue_closed_eyes: 
We should even bump to `v0.6.1`

Comment (chris-crone) [reply e2e/testdata/Dockerfile.notary-server]: latest version on the hub is 0.5.0 https://hub.docker.com/_/notary/


cc @endophage

Comment (vdemeester) [reply e2e/testdata/Dockerfile.notary-server]: ohhh :sweat:

Comment (thaJeztah) [reply e2e/testdata/Dockerfile.notary-server]: Yes, see https://github.com/docker/cli/pull/1064#discussion_r189012761

File: internal/test/output/output.go (hunk 12 above)
Comment (thaJeztah) [reply internal/test/output/output.go]: Was wondering the same when reading that; I don't think that line should disappear, will have to check why it's not printed.

Can confirm it's not there though;

```bash
mkdir foo && cd foo
echo bla > something

docker build --no-cache . -f-<<EOF
FROM alpine
COPY something /
COPY something /else/
EOF
```

```
Sending build context to Docker daemon  2.607kB
Step 1/3 : FROM alpine
 ---> 3fd9065eaf02
Step 2/3 : COPY something /
 ---> Using cache
 ---> 97e7adea2403
Step 3/3 : COPY something /else/
 ---> b0cdf35d65d6
Successfully built b0cdf35d65d6
Successfully tagged copy:latest
```

Comment (thaJeztah) [reply internal/test/output/output.go]: Same result when adding `--rm` and `--force-rm` (just in case);

```
docker build --no-cache --force-rm --rm . -f-<<EOF
FROM alpine
COPY something /
COPY something /else/
EOF
```

```
Sending build context to Docker daemon  2.607kB
Step 1/3 : FROM alpine
 ---> 3fd9065eaf02
Step 2/3 : COPY something /
 ---> cf910fa96aa7
Step 3/3 : COPY something /else/
 ---> 56bcd1f3113b
Successfully built 56bcd1f3113b
```

Comment (dnephin) [reply internal/test/output/output.go]: Instead of making the assertion less strict (and less obvious) I would prefer to have different expected values based on the version.

Comment (chris-crone) [reply internal/test/output/output.go]: @thaJeztah Thanks for the investigation. Chatting to @simonferquel offline; he believes the missing line is a bug.

@dnephin My thinking was that we should just be looking for lines of interest in order and that was enough of a test. I agree with you that it makes the check less clear. Since this appears to be an actual bug; I will wait for it to be fixed and then remove the check function.

Comment (chris-crone) [reply internal/test/output/output.go]: Created a moby issue for this: https://github.com/moby/moby/issues/37164

Updated PR to remove the addition of `AssertContains`.

File: internal/test/environment/testenv.go [hunk 14]
```diff
@@ -14,7 +15,63 @@ func Setup() error {
...
+		}
+	}
+
+	if val := boolFromString(os.Getenv("TEST_REMOTE_DAEMON")); val {
+		if err := os.Setenv("REMOTE_DAEMON", "1"); err != nil {
+			return err
+		}
+	}
+
+	if val := boolFromString(os.Getenv("TEST_SKIP_PLUGIN_TESTS")); val {
```
Comment (silvin-lubecki) [inline internal/test/environment/testenv.go]: There's like a pattern here 😇, I think it could be factorized, at least between `TEST_KUBECONFIG`, `TEST_REMOTE_DAEMON` and `TEST_SKIP_PLUGIN_TESTS`

File: e2e/stack/deploy_test.go [hunk 15]
```diff
@@ -1,25 +1,43 @@
...
+	"github.com/gotestyourself/gotestyourself/skip"
 )
 
 func TestDeployWithNamedResources(t *testing.T) {
-	stackname := "test-stack-deploy-with-names"
+	t.Run("Swarm", func(t *testing.T) {
+		testDeployWithNamedResources(t, "swarm")
+	})
+	t.Run("Kubernetes", func(t *testing.T) {
+		// FIXME(chris-crone): currently does not work with compose for kubernetes.
```
Comment (silvin-lubecki) [inline e2e/stack/deploy_test.go]: I'll try to see what's wrong here.

File: e2e/stack/remove_test.go [hunk 16]
```diff
@@ -1,57 +1,77 @@
...
 	}
 }
 
-func taskCount(stackname string, expected int) func(t poll.LogT) poll.Result {
+func taskCount(orchestrator, stackname string, expected int) func(t poll.LogT) poll.Result {
 	return func(poll.LogT) poll.Result {
-		result := icmd.RunCommand(
-			"docker", "stack", "ps", "-f=desired-state=running", stackname)
+		args := []string{"--orchestrator", orchestrator, "stack", "ps", stackname}
+		// FIXME(chris-crone): remove when filtering is supported with kubernetes.
```
[the remark this answers is not in the document]
Comment (chris-crone) [reply e2e/stack/remove_test.go]: Unfortunately not for desired state
